core/src/main/java/io/stargate/core/activator/BaseActivator.java

```diff
@@ -0,0 +1,169 @@
+            dependentService.getName());
+        context.addServiceListener(
+            this, String.format("(objectClass=%s)", dependentService.getName()));
+      } else {
+        // if the service is present, get it
+        registeredServices.put(dependentService, context.getService(serviceReference));
+      }
+    }
+
+    if (registeredServices.values().stream().anyMatch(Objects::isNull)) {
+      // It will be started once all dependent services are registered
+      return;
+    }
+
+    // all dependant services are present
+    startServiceInternal(new LinkedList<>(registeredServices.values()));
+  }
+
+  @Override
+  public synchronized void stop(BundleContext context) throws Exception {
+    // no-op
+  }
+
+  /**
+   * It will try to match all dependentServices with a ServiceEvent notification. If the
+   * notification for dependant service contains a non-null value, it will start the target service.
+   * After the notification is handled, it checks if all dependentServices are not null. If they
+   * are, the client's provided {@link this#createService(List)} is called, and the service is
+   * registered. It will not register the service if it was already registered in the {@link
+   * this#start(BundleContext)}.
+   */
+  @Override
+  public synchronized void serviceChanged(ServiceEvent event) {
+    for (Class<?> dependentService : dependentServices) {
+      Object registeredService = BundleUtils.getRegisteredService(context, event, dependentService);
+      if (registeredService != null) {
+        // capture registration only if the instance is not null
+        registeredServices.put(dependentService, registeredService);
+      }
+    }
+    if (registeredServices.values().stream().allMatch(Objects::nonNull)) {
+      startServiceInternal(new LinkedList<>(registeredServices.values()));
+    }
+  }
+
+  private synchronized void startServiceInternal(List<Object> dependentServices) {
+    if (started) {
+      logger.info("The {} is already started. Ignoring the start request.", activatorName);
+      return;
+    }
+    started = true;
+    ServiceAndProperties service = createService(dependentServices);
+    context.registerService(targetServiceClass.getName(), service.service, service.properties);
+    logger.info("Started {}....", activatorName);
+  }
+
+  /**
+   * Clients should override this method to create the Service that will be registered in the OSGi
+   * container. The dependant services will contain all services registered passed to the
+   * constructor of this class as {@code List<Class<?>> dependentServices}. The ordering will be the
+   * same. You can safely cast the objects to the expected types according to the dependentServices.
+   *
+   * @return ServiceAndProperties that has the service for OSGi registration and the properties that
+   *     will be passed.
+   */
+  protected abstract ServiceAndProperties createService(List<Object> dependentServices);
```

Comment:
Could we provide type safe references to dependencies? One possible way is to take "service selector" objects in the constructor and populate them when the services become available.

Reply:
I don't see a way of making it more compile-time type-safe without reducing the extensibility of this base class. 
I think that at some point we need to operate on `Object`.
Could you elaborate more (give example) how would you make it type-safe?

Reply:
Rough sketch:

```

class BaseActivator {
protected abstract Collection<ServicePointer<?>> dependencies();
... 
}

class ServicePointer<T> {

  ServicePointer(Class<T> targetClass, String filter) {...}

  // called by BaseActivator
  setService(T service) {...}

 // called by user (sub-class of BaseActivator)
T getService() {...}
}
```

Reply:
at some point, you will need to invoke `  setService(T service) {...}` that takes T, but the service retrieved here:
`public Object addingService(ServiceReference<Object> ref)`
is of an `Object` type, so it is impossible to call `setService(T)`.

Second issue - the dependencies must be of a `?` type. Because of that, when the sub-class of `BaseActivator` will get it, it is still impossible to have the type-safety. You will still need to cast:
```
@Override
  protected ServiceAndProperties createService(List<ServicePointer<?>> dependentServices) {
	  Object service = dependentServices.get(0).getService(); // need to cast
            ...
```
So I don't see how the ServicePointer will solve this issue.

Reply:
Yes, I think it's unavoidable in the java type system. However, I think it is preferable to have the base / utility class deal with type casts so that sub-classes (majority) will not have to do type casts. 

I think we can also check the actual type of the service object before injecting it into `ServicePointer` since we that the `Class` objects for its expected type (e.g. use `Class.cast(...)`).

Reply:
The check before injecting already exists:
```
  for (Map.Entry<Class<?>, Object> registeredService : registeredServices.entrySet()) {
        if (registeredService.getKey().isAssignableFrom(service.getClass())) {
          logger.debug("{} using service: {}", activatorName, ref.getBundle());
          registeredServices.put(registeredService.getKey(), service);
        }
      }
```

Regarding hiding it in the parent class, of course, it would be better. Unfortunately, even with your approach, sub-classes will need to cast. The reason for this is the wildcard `?`:
```
Collection<ServicePointer<?>> dependencies();
```
as long as you have one service pointer, it will work:
```
ServicePointer<String> stringServicePointer = new ServicePointer<>(String.class);
String service = stringServicePointer.getService(); // works
```
but please keep in mind that it will be a list of different `T` so for example:
```
List<ServicePointer<?>> dependencies = Arrays.asList(
				new ServicePointer<>(String.class),
				new ServicePointer<>(Integer.class)
		);
```
event if every single `ServicePointer` is type-safe the caller will still operate on `Object`:
```
		Object service1 = dependencies.get(0).getService();
		Object service2 = dependencies.get(0).getService();
```
and do the cast. Using `cast` does not solve this issue:
```
Object cast = dependencies.get(0).targetClass.cast(dependencies.get(0).getService());
```
So as long as the client needs to know exact types (and those are all our cases), the cast MUST be in the client code. wdyt?

Reply:
the `?` is only at the base class level. Sub-classes will not their types, e.g.

```
class ExampleServiceActivator extends BaseActivator {

  private final ServiceReference<Persistence> persistence = new ...

 @Override
 protected Collection<ServiceReference<?>> dependencies() {
   // this may need to be adjusted... not sure about exact signature, but should work in general, I believe
   return Collections.singletonList(persistence)
 }

 @Override
 void startServioce() { // or whatever 
   MyService = new MyService(persistence.get());
 }
}

class MyService {
 MyService(Persistence p) {...}
}
```

Reply:
That was a very good idea! Thank you. I've implemented the type-safe variant of this API using your approach, please take a look.